tests/nnet/losses/test_focal_loss.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+from numpy.testing import assert_allclose
+from hypothesis import given, assume
+import hypothesis.strategies as st
+import hypothesis.extra.numpy as hnp
+import pytest
+
+from mygrad import Tensor, sum, log
+from mygrad.nnet.activations import softmax
+from mygrad.nnet.losses import softmax_focal_loss, focal_loss
+
+
+@pytest.mark.parametrize("fn", (softmax_focal_loss, focal_loss))
+@pytest.mark.parametrize(
+    ("data", "labels"),
+    [
+        (np.ones((2,), dtype=float), np.zeros((2,), dtype=int)),  # 1D data
+        (np.ones((2, 1), dtype=float), np.zeros((2,), dtype=float)),  # non-int labels
+        (np.ones((2, 1), dtype=float), np.zeros((2, 1), dtype=int)),  # bad label-ndim
+        (np.ones((2, 1), dtype=float), np.zeros((3,), dtype=int)),  # bad label-shape
+    ],
+)
+def test_input_validation(fn, data, labels):
+    with pytest.raises((ValueError, TypeError)):
+        fn(data, labels)
+
+
+@given(
+    num_datum=st.integers(1, 100),
+    num_classes=st.integers(1, 15),
+    alpha=st.floats(-1, 1),
+    gamma=st.floats(0, 5),
+    data=st.data(),
+    grad=st.floats(-1, 1),
+    target_type=st.sampled_from((np.array, Tensor)),
+)
+def test_softmax_focal_loss(num_datum, num_classes, alpha, gamma, data, grad, target_type):
```

Comment:
I think I will want to loop back to this and write tests using uber, since that makes nice extra checks; like protects against input mutation, propagation of `constant` etc. Not necessary for this PR but I'll make an issue about this.